### PR TITLE
Add helpers to create errdef errors

### PR DESCRIPTION
**Pull request by cpuguy83**
Instead of having to create a bunch of custom error types that are doing
nothing but wrapping another error in sub-packages, use a common helper
to create errors of the requested type.

e.g. instead of re-implementing this over and over:

```go
type notFoundError struct {
  cause error
}

func(e notFoundError) Error() string {
  return e.cause.Error()
}

func(e notFoundError) NotFound() {}

func(e notFoundError) Cause() error {
  return e.cause
}
```

Packages can instead just do:

```
  errdefs.NotFound(err)
```

### CODE REVIEW

**Comment by dnephin** [comment]
I think you forgot to `git add` a file

File: daemon/archive.go
Comment (thaJeztah) [inline daemon/archive.go]: Wondering if we're now making separation between API and internals worse?

Comment (cpuguy83) [reply daemon/archive.go]: This is the daemon package.

**Comment by cpuguy83** [comment]
Ping this should be good to go.

File: api/server/httputils/errors.go
Comment (dnephin) [inline api/server/httputils/errors.go]: `IsAlreadyExist()` seems like it could mean not modified, but in other contexts it could mean Conflict.

File: api/errdefs/helpers.go
Comment (dnephin) [inline api/errdefs/helpers.go]: This is the opposite of `containerd`, which uses instances of an error as the root cause, and then wraps with more context. That's probably ok, but something we should be careful of when converting `containerd` errors. I guess this way requires `getImplementer()`, since `errors.Cause()` will not return the correct error (another thing we need to be watch out for).

File: daemon/logs.go
Comment (dnephin) [inline daemon/logs.go]: Should this be `Conflict()` (I believe others were coverted to Conflict)

File: api/errdefs/helpers.go
Comment (cpuguy83) [reply api/errdefs/helpers.go]: errdefs here is implemented slightly differently from containerd, where containerd has a set of base errors like `ErrNotFound = errors.New("not found")` which you wrap and then compare against the base error like `if errors.Cause(err) == ErrNotFound`.
But yeah we need to be careful any time we are dealing with errors from other repos.

File: daemon/logs.go
Comment (cpuguy83) [reply daemon/logs.go]: Yes should be conflict.

File: api/server/httputils/errors.go
Comment (cpuguy83) [reply api/server/httputils/errors.go]: Agreed, but in that scenario it should be up to the thing that received the error to determine if it's a hard failure like a conflict or already exists.
The doc string for `ErrAlreadyExists` spells out that this is a special case of `ErrNotModified`

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
I very much like the changes in this PR (and the helpers), but I'm a bit on the fence; trying to phrase my thoughts, but let me know if they make sense.

The initial goal of these "standardized error interfaces" was so that the docker "internals" provide more information about an error (allowing an API to consume the additional information), but don't have to be aware of API-specifics (more specifically: don't assume a HTTP api, and don't let HTTP status codes find their way into the internals).

The error interfaces definitely helped realizing that, and this code makes it easier to use them.

However, both the _definition_, _conversion_ (of standard error-types to HTTP status codes), and now the _helpers_ are included in the API code. I'm not sure they _should be_, because now the "internal" code depends on the API, which (imo) it shouldn't be (there are already other parts of the API that leak into internals, and that should probably also be looked into).

Ideally, it would be possible to remove the `api`, swap it for a different implementation, and build something from that; with the errdefs being part of the API, that's not possible.

So; I'm wondering; should we move the `errdefs` to a package/directory _separate_ from the API? Both the daemon/internals and API can use that package, but it wouldn't be _tied_ to the API.

**Comment by cpuguy83** [comment]
The errdefs are not part of the public API, they are a common API to use for errors internally.
The conversion to http is handled separately from the definition of the errors, and even likely in the future we could further annotate the error on the wire with more details so that they can be decoded properly on the other side (e.g. the errcode package in distribution).

Basically, these are not part of the API implementation which can easily be swapped out and not affect internal code.

**Comment by thaJeztah** [comment]
Perhaps it's just me being OCD, and focussing on it being inside the `api/` directory; could they live inside `pkg/`?

**Comment by cpuguy83** [comment]
I very much see these as part of the internal API and docker specific.... but we could move to the root of the repo.

**Comment by dnephin** [comment]
`errdefs` at the root SGTM

**Comment by thaJeztah** [comment]
> I very much see these as part of the internal API and docker specific

They felt quite generic to me, but yes, we can keep them out of `pkg` to allow us to have errors that are specific to our use.

>  but we could move to the root of the repo.

That would work for me; do you want to discuss this more with other maintainers? If so, I can add it to the maintainers meeting for a quick chat (let me know if you want that)

**Comment by cpuguy83** [comment]
I've moved the package from `api/errdefs` to just `errdefs`.

File: api/server/router/network/network_routes.go
Comment (thaJeztah) [inline api/server/router/network/network_routes.go]: Same warning does not apply here I assume? (re: `libnetwork.ErrNoSuchNetwork()` being special)

File: daemon/stop.go
Comment (thaJeztah) [inline daemon/stop.go]: Same question here (re: reversing order of wrap)

File: errdefs/defs.go
Comment (thaJeztah) [inline errdefs/defs.go]: Saw @dnephin already commented on this type https://github.com/moby/moby/pull/35638/files#r160793335, just wanted to add that naming it `ErrAlreadyExists` "feels" like an `ErrConflict`.

`ErrAlreadyExists` and `ErrNotModified` inside the API are converted to a HTTP 304 (`http.StatusNotModified`), which is not an _error_ condition, so perhaps we should have a look at where they're used, and if it's correct that an error is returned at all.

(not for this PR, but we may want to add a TODO or issue somewhere)

File: errdefs/defs.go
Comment (thaJeztah) [inline errdefs/defs.go]: I notice there's no HTTP translation for `ErrCancelled`, `ErrDeadline` and `ErrDataLoss` in httputils/errors.go, correct?

File: errdefs/defs.go
Comment (thaJeztah) [inline errdefs/defs.go]: Would this be a `StatusRequestTimeout`? If so, wondering if it should have a more generic name (but that's a nit)

File: errdefs/defs.go
Comment (thaJeztah) [inline errdefs/defs.go]: Looks like `ErrDataLoss` is not used anywhere


I see some mention of these but for `gRPC` errors; https://github.com/moby/moby/blob/6540d5581473c292ade46ee1b620172441943f4b/api/server/httputils/errors.go#L122-L128

File: libcontainerd/errors.go
Comment (thaJeztah) [inline libcontainerd/errors.go]: For a follow-up, we should get rid of these helpers (now it's helpers using helpers :joy:)

File: registry/auth.go
Comment (thaJeztah) [inline registry/auth.go]: This should probably be `errdefs.Unauthorized()`

File: volume/local/local.go
Comment (thaJeztah) [inline volume/local/local.go]: Looks like this needs a `wrap` to preserve the original error?

File: volume/local/local.go
Comment (thaJeztah) [inline volume/local/local.go]: Same question as elsewhere; why `wrap` order is reversed

File: volume/local/local.go
Comment (thaJeztah) [inline volume/local/local.go]: Same question as elsewhere; why `wrap` order is reversed

**Review by thaJeztah — CHANGES_REQUESTED** [review]
thanks for moving the package 🤗 

Left comments and questions inline, but overall looking good 👍

File: api/server/router/network/network_routes.go
Comment (cpuguy83) [reply api/server/router/network/network_routes.go]: It doesn't seem like it since nothing should be consuming this other than the HTTP API.
In any case the caller can use `errors.Cause()`

File: daemon/exec.go
Comment (cpuguy83) [reply daemon/exec.go]: 🤷‍♂️ I like it better.

File: errdefs/defs.go
Comment (cpuguy83) [reply errdefs/defs.go]: `Request` of what?
This is really for propagating context errors.

File: errdefs/defs.go
Comment (cpuguy83) [reply errdefs/defs.go]: I plan to :)
e.g. for corrupted json log entries.

File: registry/auth.go
Comment (cpuguy83) [reply registry/auth.go]: I don't know why I chose this, but I think you are right.

File: errdefs/defs.go
Comment (cpuguy83) [reply errdefs/defs.go]: Correct, it's just a 500.

File: errdefs/defs.go
Comment (cpuguy83) [reply errdefs/defs.go]: Indeed we could switch it... it may make since to do this.

Comment (cpuguy83) [reply errdefs/defs.go]: "Make sense" because it's likely safer to return a conflict than a not modified.

File: errdefs/defs.go
Comment (thaJeztah) [reply errdefs/defs.go]: nit: perhaps make it explicit, so that it's clear it's a deliberate choice to make them 500's, and not because we "forgot"

Comment (thaJeztah) [reply errdefs/defs.go]: Oh, never mind; didn't look at the code yet; I see you added them :heart:

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
